<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="spicyWorkflow">
<title>Performing a spatial analysis of multiplexed tissue imaging data. • spicyWorkflow</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Performing a spatial analysis of multiplexed tissue imaging data.">
<meta property="og:description" content="spicyWorkflow">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">spicyWorkflow</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/spicyWorkflow.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/SydneyBioX/spicyWorkflow/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Performing a spatial analysis of multiplexed tissue imaging data.</h1>
                        <h4 data-toc-skip class="author">Nicholas
Robertson</h4>
            <address class="author_afil">
      School of Mathematics and Statistics, University of Sydney,
Australia<br><h4 data-toc-skip class="author">Alexander
Nicholls</h4>
            <address class="author_afil">
      Westmead Institute for Medical Research, University of Sydney,
Australia<br><h4 data-toc-skip class="author">Nicholas
Canete</h4>
            <address class="author_afil">
      Westmead Institute for Medical Research, University of Sydney,
Australia<br><h4 data-toc-skip class="author">Elijah
Willie</h4>
            <address class="author_afil">
      Westmead Institute for Medical Research, University of Sydney,
AustraliaSchool of Mathematics and Statistics, University of Sydney,
Australia<br><h4 data-toc-skip class="author">Ellis
Patrick</h4>
            <address class="author_afil">
      Westmead Institute for Medical Research, University of Sydney,
AustraliaSchool of Mathematics and Statistics, University of Sydney,
Australia<br><h4 data-toc-skip class="date">27 July, 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/SydneyBioX/spicyWorkflow/blob/HEAD/vignettes/spicyWorkflow.Rmd" class="external-link"><code>vignettes/spicyWorkflow.Rmd</code></a></small>
      <div class="d-none name"><code>spicyWorkflow.Rmd</code></div>
    </address>
</address>
</address>
</address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="version-info">Version Info<a class="anchor" aria-label="anchor" href="#version-info"></a>
</h2>
<p>
</p>
<p><strong>R version</strong>: R version 4.2.1 (2022-06-23) <br><strong>Bioconductor version</strong>: 3.16 <br></p>
<p>Understanding the interplay between different types of cells and
their immediate environment is critical for understanding the mechanisms
of cells themselves and their function in the context of human diseases.
Recent advances in high dimensional in situ cytometry technologies have
fundamentally revolutionized our ability to observe these complex
cellular relationships providing an unprecedented characterisation of
cellular heterogeneity in a tissue environment.</p>
<p>We have developed an analytical framework for analysing data from
high dimensional in situ cytometry assays including CODEX, CycIF, IMC
and High Definition Spatial Transcriptomics. Implemented in R, this
framework makes use of functionality from our Bioconductor packages
spicyR, lisaClust, scFeatures, treekoR, FuseSOM, simpleSeg and
ClassifyR. Below we will provide an overview of key steps which are
needed to interrogate the comprehensive spatial information generated by
these exciting new technologies including cell segmentation, feature
normalisation, cell type identification, microenvironment
characterisation, spatial hypothesis testing and patient classification.
Ultimately, our modular analysis framework provides a cohesive and
accessible entry point into spatially resolved single cell data analysis
for any R-based bioinformaticians.</p>
</div>
<div class="section level2">
<h2 id="loading-r-packages">Loading R packages<a class="anchor" aria-label="anchor" href="#loading-r-packages"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BodenmillerGroup/cytomapper" class="external-link">cytomapper</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/SydneyBioX/simpleSeg" class="external-link">simpleSeg</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">FuseSOM</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/" class="external-link">ggpubr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">spicyR</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ClassifyR</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">lisaClust</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="global-paramaters">Global paramaters<a class="anchor" aria-label="anchor" href="#global-paramaters"></a>
</h2>
<p>It is convenient to set the number of cores for running code in
parallel. Please choose a number that is appropriate for your
resources.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nCores</span> <span class="op">&lt;-</span> <span class="fl">20</span></span>
<span><span class="va">BPPARAM</span> <span class="op">&lt;-</span> <span class="fu">simpleSeg</span><span class="fu">:::</span><span class="fu"><a href="https://rdrr.io/pkg/simpleSeg/man/generateBPParam.html" class="external-link">generateBPParam</a></span><span class="op">(</span><span class="va">nCores</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="context">Context<a class="anchor" aria-label="anchor" href="#context"></a>
</h2>
<p>In the following we will reanalyse some MIBI-TOF data <a href="https://www.sciencedirect.com/science/article/pii/S0092867421014860?via%3Dihub#!" class="external-link">(Risom
et al, 2022)</a> profiling the spatial landscape of ductal carcinoma in
situ (DCIS), which is a pre-invasive lesion that is thought to be a
precursor to invasive breast cancer (IBC). The key conclusion of this
manuscript (amongst others) is that spatial information about cells can
be used to predict disease progression in patients. We will use our
spicy workflow to make a similar conclusion.</p>
<p>The R code for this analysis is available on github <a href="https://github.com/SydneyBioX/spicyWorkflow" class="external-link">https://github.com/SydneyBioX/spicyWorkflow</a>.
A mildly <a href="https://github.com/SydneyBioX/spicyWorkflow/blob/master/organisePublishedData.R" class="external-link">processed</a>
version of the data used in the manuscript is available in this
repository.</p>
</div>
<div class="section level2">
<h2 id="read-in-images">Read in images<a class="anchor" aria-label="anchor" href="#read-in-images"></a>
</h2>
<p>The images are stored in the <code>images</code> folder within the
<code>data</code> folder. Here we use <code>readImages()</code> from the
<code>EBImage</code> package to read these into R. If memory is a
restricting factor, and the files are in a slightly different format,
you could use <code><a href="https://rdrr.io/pkg/cytomapper/man/loadImages.html" class="external-link">loadImages()</a></code> from the <code>cytomapper</code>
package to load all the tiff images into a <code>CytoImageList</code>
object, which can store the images as h5 on-disk.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pathToImages</span> <span class="op">&lt;-</span> <span class="st">"../inst/extdata/images"</span></span>
<span></span>
<span><span class="co"># Get directories of images</span></span>
<span><span class="va">imageDirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="va">pathToImages</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">imageDirs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="va">pathToImages</span>, full.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get files in each directory</span></span>
<span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">imageDirs</span>, <span class="va">list.files</span>, pattern <span class="op">=</span> <span class="st">"tif"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span>, simplify <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Read files with readImage from EBImage</span></span>
<span><span class="va">images</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">files</span>, <span class="fu">EBImage</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/EBImage/man/io.html" class="external-link">readImage</a></span>, as.is <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>We will make use of the <code>on_disk</code> option to convert our
images to a <code>CytoImageList</code> with the images not held in
memory.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Store images in a CytoImageList with images on_disk as h5 files to save memory. </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"../inst/extdata/h5Files"</span><span class="op">)</span></span>
<span><span class="va">images</span> <span class="op">&lt;-</span> <span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/CytoImageList.html" class="external-link">CytoImageList</a></span><span class="op">(</span><span class="va">images</span>, </span>
<span>                                    on_disk <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                    h5FilesPath <span class="op">=</span> <span class="st">"../inst/extdata/h5Files"</span>, </span>
<span>                                    BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/gc.html" class="external-link">gc</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            used  (Mb) gc trigger    (Mb)   max used    (Mb)</span></span>
<span><span class="co">## Ncells 11172225 596.7   19785521  1056.7   14792111   790.0</span></span>
<span><span class="co">## Vcells 18263177 139.4 1617401760 12339.8 2017382849 15391.5</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="load-the-clinical-data">Load the clinical data<a class="anchor" aria-label="anchor" href="#load-the-clinical-data"></a>
</h2>
<p>To associate features in our image with disease progression, it is
important to read in information which links image identifiers to their
progression status. We will do this here, making sure that our
<code>imageID</code> match. ## Read the clinical data</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in clinical data, manipulate imageID and select columns</span></span>
<span><span class="va">clinical</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"../inst/extdata/1-s2.0-S0092867421014860-mmc1.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">clinical</span> <span class="op">&lt;-</span> <span class="va">clinical</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>imageID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Point"</span>, <span class="va">PointNumber</span>, <span class="st">"_pt"</span>, <span class="va">Patient_ID</span>, <span class="st">"_"</span>, <span class="va">TMAD_Patient</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">clinical</span><span class="op">$</span><span class="va">imageID</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"normal"</span>, <span class="va">clinical</span><span class="op">$</span><span class="va">Tissue_Type</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">$</span><span class="va">imageID</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"normal"</span>, <span class="va">clinical</span><span class="op">$</span><span class="va">Tissue_Type</span><span class="op">)</span><span class="op">]</span>, <span class="st">"_Normal"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">clinicalVariables</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"imageID"</span>, <span class="st">"Patient_ID"</span>,<span class="st">"Status"</span>, <span class="st">"Age"</span>, <span class="st">"SUBTYPE"</span>, <span class="st">"PAM50"</span>, <span class="st">"Treatment"</span>, <span class="st">"DCIS_grade"</span>, <span class="st">"Necrosis"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">clinical</span><span class="op">$</span><span class="va">imageID</span></span></code></pre></div>
<div class="section level3">
<h3 id="put-the-clinical-data-into-the-coldata-of-singlecellexperiment">Put the clinical data into the colData of SingleCellExperiment<a class="anchor" aria-label="anchor" href="#put-the-clinical-data-into-the-coldata-of-singlecellexperiment"></a>
</h3>
<p>We can then store the clinical information in the <code>mcols</code>
of the <code>CytoImageList</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add the clinical data to mcols of images.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">clinical</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span>, <span class="va">clinicalVariables</span><span class="op">]</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="simpleseg-segment-the-cells-in-the-images">SimpleSeg: Segment the cells in the images<a class="anchor" aria-label="anchor" href="#simpleseg-segment-the-cells-in-the-images"></a>
</h2>
<p>Our simpleSeg R package on <a href="https://github.com/SydneyBioX/simpleSeg" class="external-link">https://github.com/SydneyBioX/simpleSeg</a>
provides a series of functions to generate simple segmentation masks of
images. These functions leverage the functionality of the <a href="https://bioconductor.org/packages/release/bioc/vignettes/EBImage/inst/doc/EBImage-introduction.html" class="external-link">EBImage</a>
package on Bioconductor. For more flexibility when performing your
segmentation in R we recommend learning to use the EBimage package. A
key strength of the simpleSeg package is that we have coded multiple
ways to perform some simple segmentation operations as well as
incorporating multiple automatic procedures to optimise some key
parameters when these aren’t specified.</p>
<div class="section level3">
<h3 id="run-simpleseg">Run simpleSeg<a class="anchor" aria-label="anchor" href="#run-simpleseg"></a>
</h3>
<p>If your images are stored in a <code>list</code> or
<code>CytoImageList</code> they can be segmented with a simple call to
<code><a href="https://rdrr.io/pkg/simpleSeg/man/simpleSeg.html" class="external-link">simpleSeg()</a></code>. Here we have ask <code>simpleSeg</code> to do
multiple things. First, we would like to use a combination of principal
component analysis of all channels guided by the H33 channel to
summarised the nuclei signal in the images. Secondly, to estimate the
cell body of the cells we will simply dilate out from the nuclei by 2
pixels. We have also requested that the channels be square root
transformed and that a minimum cell size of 40 pixels be used as a size
selection step.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate segmentation masks </span></span>
<span><span class="va">masks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/simpleSeg/man/simpleSeg.html" class="external-link">simpleSeg</a></span><span class="op">(</span><span class="va">images</span>,</span>
<span>                   nucleus <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PCA"</span>, <span class="st">"HH3"</span><span class="op">)</span>,</span>
<span>                   cellBody <span class="op">=</span> <span class="st">"dilate"</span>,</span>
<span>                   transform <span class="op">=</span> <span class="st">"sqrt"</span>,</span>
<span>                   sizeSelection <span class="op">=</span> <span class="fl">40</span>,</span>
<span>                   discSize <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                   cores <span class="op">=</span> <span class="va">nCores</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualise-separation">Visualise separation<a class="anchor" aria-label="anchor" href="#visualise-separation"></a>
</h3>
<p>The <code>display</code> and <code>colorLabels</code> functions in
<code>EBImage</code> make it very easy to examine the performance of the
cell segmentation. The great thing about <code>display</code> is that if
used in an interactive session it is very easy to zoom in and out of the
image.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise segmentation performance one way.</span></span>
<span><span class="fu">EBImage</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/colorLabels.html" class="external-link">colorLabels</a></span><span class="op">(</span><span class="va">masks</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="visualise-outlines">Visualise outlines<a class="anchor" aria-label="anchor" href="#visualise-outlines"></a>
</h3>
<p>The <code>plotPixels</code> function in <code>cytomapper</code> make
it easy to overlay the masks on top of the intensities of 6 markers.
Here we can see that the segmentation appears to be performing
reasonably.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise segmentation performance another way.</span></span>
<span><span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/plotPixels.html" class="external-link">plotPixels</a></span><span class="op">(</span>image <span class="op">=</span> <span class="va">images</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, </span>
<span>                       mask <span class="op">=</span> <span class="va">masks</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, </span>
<span>                       img_id <span class="op">=</span> <span class="st">"imageID"</span>, </span>
<span>                       colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PanKRT"</span>, <span class="st">"GLUT1"</span>, <span class="st">"HH3"</span>, <span class="st">"CD3"</span>, <span class="st">"CD20"</span><span class="op">)</span>, </span>
<span>                       display <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>                       colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>HH3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"blue"</span><span class="op">)</span>, </span>
<span>                                     CD3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"purple"</span><span class="op">)</span>,</span>
<span>                                     CD20 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"green"</span><span class="op">)</span>,</span>
<span>                                     GLUT1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span>,</span>
<span>                                     PanKRT <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"yellow"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                       bcg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>HH3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1.5</span><span class="op">)</span>, </span>
<span>                                     CD3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1.5</span><span class="op">)</span>,</span>
<span>                                     CD20 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1.5</span><span class="op">)</span>,</span>
<span>                                     GLUT1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1.5</span><span class="op">)</span>,</span>
<span>                                     PanKRT <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1.5</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                       legend <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="summarise-cell-features-">Summarise cell features.<a class="anchor" aria-label="anchor" href="#summarise-cell-features-"></a>
</h2>
<p>In order to charactise the phenotypes of each of the segmented cells,
<code>measureObjects</code> from <code>cytomapper</code> will calculate
the average intensity of each channel within each cell as well as a few
morphological features. The channel intensities will be stored in the
<code>counts assay</code> in a <code>SingleCellExperiment</code>.
Information on the spatial location of each cell is stored in
<code>colData</code> in the <code>m.cx</code> and <code>m.cy</code>
columns. In addition to this, it will propagate the information we have
store in the <code>mcols</code> of our <code>CytoImageList</code> in the
<code>colData</code> of the resulting
<code>SingleCellExperiment</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Summarise the experssion of each marker in each cell</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/measureObjects.html" class="external-link">measureObjects</a></span><span class="op">(</span><span class="va">masks</span>, </span>
<span>                                    <span class="va">images</span>, </span>
<span>                                    img_id <span class="op">=</span> <span class="st">"imageID"</span>, </span>
<span>                                    BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="normalize-data">Normalize data<a class="anchor" aria-label="anchor" href="#normalize-data"></a>
</h2>
<p>We should check to see if the marker intensities of each cell require
some form of transformation or normalisation. Here we extract the
intensities from the <code>counts</code> assay. Looking at PanKRT which
should be expressed in the majority of the tumour cells, the intensities
are clearly very skewed.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract marker data and bind with information about images</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">cells</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plots densities of PanKRT for each image.</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PanKRT</span>, colour <span class="op">=</span> <span class="va">imageID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-12-1.png" width="480"></p>
<p>We can transform and normalise our data using the
<code>normalizeCells</code> function. Here we have taken the intensities
from the <code>counts</code> assay, performed a square root transform,
then for each image trimmed the 99 quantile and min-max scaled to 0-1.
This modified data is then store in the <code>norm</code> assay by
default. We can see that this normalised data appears more bimodal, not
perfect, but likely sufficient for clustering.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Transform and normalise the marker expression of each cell type.</span></span>
<span><span class="co"># Use a square root transform, then trimmed the 99 quantile </span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/simpleSeg/man/normalizeCells.html" class="external-link">normalizeCells</a></span><span class="op">(</span><span class="va">cells</span>, </span>
<span>                        transformation <span class="op">=</span> <span class="st">"asinh"</span>, </span>
<span>                        method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"trim99"</span>, <span class="st">"minMax"</span><span class="op">)</span>, </span>
<span>                        assayIn <span class="op">=</span> <span class="st">"counts"</span>, </span>
<span>                        cores <span class="op">=</span> <span class="va">nCores</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract normalised marker information.</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">cells</span>, <span class="st">"norm"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plots densities of normalised PanKRT for each image.</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">CD68</span>, colour <span class="op">=</span> <span class="va">imageID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-13-1.png" width="480"></p>
</div>
<div class="section level2">
<h2 id="fusesom-cluster-cells-into-cell-types">FuseSOM: Cluster cells into cell types<a class="anchor" aria-label="anchor" href="#fusesom-cluster-cells-into-cell-types"></a>
</h2>
<p>Our FuseSOM R package on <a href="https://github.com/ecool50/FuseSOM" class="external-link">https://github.com/ecool50/FuseSOM</a>
and provides a pipeline for the clustering of highly multiplexed in situ
imaging cytometry assays. This pipeline uses the Self Organizing Map
architecture coupled with Multiview hierarchical clustering and provides
functions for the estimation of the number of clusters.</p>
<p>Here we cluster using the <code>runFuseSOM</code> function. We have
chosen to specify the same subset of markers used in the original
manuscript for gating cell types. We have also specified the number of
clusters to identify to be <code>numClusters = 19</code>. In addtion to
this, while FuseSOM can automatically estimate a grid size for the self
organising map, we have set the grid size to be <code>size = 10</code>
to specify a 10 by 10 grid that should allow for smaller cell subsets to
be identified.</p>
<div class="section level3">
<h3 id="perform-the-clustering">Perform the clustering<a class="anchor" aria-label="anchor" href="#perform-the-clustering"></a>
</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The markers used in the original publication to gate cell types.</span></span>
<span><span class="va">useMarkers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PanKRT"</span>, <span class="st">"ECAD"</span>, <span class="st">"CK7"</span>, <span class="st">"VIM"</span>, <span class="st">"FAP"</span>, <span class="st">"CD31"</span>, <span class="st">"CK5"</span>, <span class="st">"SMA"</span>, </span>
<span>                <span class="st">"CD45"</span>, <span class="st">"CD4"</span>, <span class="st">"CD3"</span>, <span class="st">"CD8"</span>, <span class="st">"CD20"</span>, <span class="st">"CD68"</span>, <span class="st">"CD14"</span>, <span class="st">"CD11c"</span>, </span>
<span>                <span class="st">"HLADRDPDQ"</span>, <span class="st">"MPO"</span>, <span class="st">"Tryptase"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set seed.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate SOM and cluster cells into 20 groups.</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/runFuseSOM.html" class="external-link">runFuseSOM</a></span><span class="op">(</span><span class="va">cells</span>,</span>
<span>                    markers <span class="op">=</span> <span class="va">useMarkers</span>,</span>
<span>                    assay <span class="op">=</span> <span class="st">'norm'</span>,</span>
<span>                    numClusters <span class="op">=</span> <span class="fl">19</span>,</span>
<span>                    size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="attempt-to-interpret-the-phenotype-of-each-cluster">Attempt to interpret the phenotype of each cluster<a class="anchor" aria-label="anchor" href="#attempt-to-interpret-the-phenotype-of-each-cluster"></a>
</h3>
<p>We can begin the process of understanding what each of these cell
clusters are by using the <code>plotGroupedHeatmap</code> function from
<code>scater</code>. At the least, here we can see we capture all the
major immune populations that we expect to see.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise marker expression in each cluster.</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotGroupedHeatmap.html" class="external-link">plotGroupedHeatmap</a></span><span class="op">(</span><span class="va">cells</span>, </span>
<span>                           features <span class="op">=</span> <span class="va">useMarkers</span>, </span>
<span>                           group <span class="op">=</span> <span class="st">"clusters"</span>, </span>
<span>                           exprs_values <span class="op">=</span> <span class="st">"norm"</span>,</span>
<span>                           center <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                           scale <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                           zlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>,<span class="fl">3</span><span class="op">)</span>,</span>
<span>                           cluster_rows <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="check-how-many-clusters-should-be-used-">Check how many clusters should be used.<a class="anchor" aria-label="anchor" href="#check-how-many-clusters-should-be-used-"></a>
</h3>
<p>We can check to see how reasonable our choice of 20 clusters is using
the <code>estimateNumCluster</code> and the <code>optiPlot</code>
functions. Here we examine the Gap method, others such as Silhouette and
Within Cluster Distance are also available.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate metrics for estimating the number of clusters.</span></span>
<span><span class="co"># As I've already run runFuseSOM I don't need to run generateSOM().</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/estimateNumCluster.html" class="external-link">estimateNumCluster</a></span><span class="op">(</span><span class="va">cells</span>, kSeq <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/optiPlot.html" class="external-link">optiPlot</a></span><span class="op">(</span><span class="va">cells</span>, method <span class="op">=</span> <span class="st">"gap"</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="check-cluster-frequencies">Check cluster frequencies<a class="anchor" aria-label="anchor" href="#check-cluster-frequencies"></a>
</h3>
<p>We find it always useful to check the number of cells in each
cluster. Here we can see that cluster 3 is contains lots of (most likely
tumour) cells and cluster 7 contains very few cells.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check cluster frequencies.</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span><span class="op">$</span><span class="va">clusters</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## cluster_13  cluster_6 cluster_19 cluster_14 cluster_12 cluster_15 cluster_18 </span></span>
<span><span class="co">##        285        290        348        350        389        475        526 </span></span>
<span><span class="co">##  cluster_7 cluster_17 cluster_16 cluster_11 cluster_10  cluster_4  cluster_8 </span></span>
<span><span class="co">##        655        803       1114       1193       1267       1811       1901 </span></span>
<span><span class="co">##  cluster_9  cluster_1  cluster_2  cluster_5  cluster_3 </span></span>
<span><span class="co">##       2445       2660       4644       5069      36335</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="test-for-association-between-the-proportion-of-each-cell-type-and-progression-status">Test For association between the proportion of each cell type and
progression status<a class="anchor" aria-label="anchor" href="#test-for-association-between-the-proportion-of-each-cell-type-and-progression-status"></a>
</h2>
<p>We recommend using a package such as <code>diffcyt</code> for testing
for changes in abundance of cell types. However, the
<code>colTest</code> function allows us to quickly test for associations
between the proportions of the cell types and progression status using
either Wilcoxon rank sum tests or t-tests. Here we see a p-value less
than 0.05, but this does not equate to a small FDR.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Select cells which belong to individuals with progressor status.</span></span>
<span><span class="va">cellsToUse</span> <span class="op">&lt;-</span> <span class="va">cells</span><span class="op">$</span><span class="va">Status</span><span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nonprogressor"</span>, <span class="st">"progressor"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform simple wicoxon rank sum tests on the columns of the proportion matrix.</span></span>
<span><span class="va">testProp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/colTest.html" class="external-link">colTest</a></span><span class="op">(</span><span class="va">cells</span><span class="op">[</span>, <span class="va">cellsToUse</span><span class="op">]</span>, </span>
<span>                    condition <span class="op">=</span> <span class="st">"Status"</span>, </span>
<span>                    feature <span class="op">=</span> <span class="st">"clusters"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">testProp</span></span></code></pre></div>
<pre><code><span><span class="co">##              W  pval adjPval    cluster</span></span>
<span><span class="co">## cluster_8  190 0.039    0.33  cluster_8</span></span>
<span><span class="co">## cluster_12 200 0.044    0.33 cluster_12</span></span>
<span><span class="co">## cluster_5  420 0.052    0.33  cluster_5</span></span>
<span><span class="co">## cluster_10 380 0.170    0.73 cluster_10</span></span>
<span><span class="co">## cluster_11 380 0.200    0.73 cluster_11</span></span>
<span><span class="co">## cluster_19 240 0.230    0.73 cluster_19</span></span>
<span><span class="co">## cluster_7  260 0.360    0.89  cluster_7</span></span>
<span><span class="co">## cluster_16 360 0.390    0.89 cluster_16</span></span>
<span><span class="co">## cluster_1  270 0.520    0.89  cluster_1</span></span>
<span><span class="co">## cluster_6  340 0.560    0.89  cluster_6</span></span>
<span><span class="co">## cluster_9  340 0.580    0.89  cluster_9</span></span>
<span><span class="co">## cluster_2  330 0.650    0.89  cluster_2</span></span>
<span><span class="co">## cluster_4  330 0.690    0.89  cluster_4</span></span>
<span><span class="co">## cluster_15 290 0.700    0.89 cluster_15</span></span>
<span><span class="co">## cluster_14 290 0.710    0.89 cluster_14</span></span>
<span><span class="co">## cluster_3  290 0.750    0.89  cluster_3</span></span>
<span><span class="co">## cluster_13 300 0.810    0.91 cluster_13</span></span>
<span><span class="co">## cluster_17 310 0.920    0.93 cluster_17</span></span>
<span><span class="co">## cluster_18 310 0.930    0.93 cluster_18</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">imagesToUse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">)</span><span class="op">[</span><span class="va">clinical</span><span class="op">[</span>, <span class="st">"Status"</span><span class="op">]</span><span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nonprogressor"</span>, <span class="st">"progressor"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">prop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/getProp.html" class="external-link">getProp</a></span><span class="op">(</span><span class="va">cells</span>, feature <span class="op">=</span> <span class="st">"clusters"</span><span class="op">)</span></span>
<span><span class="va">clusterToUse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">testProp</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span> <span class="va">prop</span><span class="op">[</span><span class="va">imagesToUse</span>, <span class="va">clusterToUse</span><span class="op">]</span> <span class="op">~</span> <span class="va">clinical</span><span class="op">[</span><span class="va">imagesToUse</span>, <span class="st">"Status"</span><span class="op">]</span> <span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<div class="section level3">
<h3 id="dimension-reduction">Dimension reduction<a class="anchor" aria-label="anchor" href="#dimension-reduction"></a>
</h3>
<p>As our data is stored in a <code>SingleCellExperiment</code> we can
also use <code>scater</code> to perform and visualise our data in a
lower dimension to look for image or cluster differences.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span><span class="co"># Perform dimension reduction using UMP.</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">cells</span>, </span>
<span>                         subset_row <span class="op">=</span> <span class="va">useMarkers</span>, </span>
<span>                         exprs_values <span class="op">=</span> <span class="st">"norm"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select a subset of images to plot.</span></span>
<span><span class="va">someImages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">10</span>,<span class="fl">20</span>,<span class="fl">40</span>,<span class="fl">50</span>,<span class="fl">60</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># UMAP by imageID.</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">cells</span><span class="op">[</span>,<span class="fu">colData</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span><span class="op">$</span><span class="va">imageID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">someImages</span><span class="op">]</span>, dimred<span class="op">=</span><span class="st">"UMAP"</span>, colour_by<span class="op">=</span><span class="st">"imageID"</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># UMAP by cell type cluster.</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">cells</span><span class="op">[</span>,<span class="fu">colData</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span><span class="op">$</span><span class="va">imageID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">someImages</span><span class="op">]</span>, dimred<span class="op">=</span><span class="st">"UMAP"</span>, colour_by<span class="op">=</span><span class="st">"clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-19-2.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="spicyr-test-spatial-relationships">spicyR: test spatial relationships<a class="anchor" aria-label="anchor" href="#spicyr-test-spatial-relationships"></a>
</h2>
<p>Our spicyR package (<a href="https://www.bioconductor.org/packages/devel/bioc/html/spicyR.html" class="external-link uri">https://www.bioconductor.org/packages/devel/bioc/html/spicyR.html</a>)[<a href="https://www.bioconductor.org/packages/devel/bioc/html/spicyR.html" class="external-link uri">https://www.bioconductor.org/packages/devel/bioc/html/spicyR.html</a>]
provides a series of functions to aid in the analysis of both
immunofluorescence and mass cytometry imaging data as well as other
assays that can deeply phenotype individual cells and their spatial
location. Here we use the <code>spicy</code> function to test for
changes in the spatial relationships between pairwise combinations of
cells. We quantify spatial relationships using a combination of three
radii <code>Rs = c(20, 50, 100)</code> and mildly account for some
global tissue structure using <code>sigma = 50</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test for changes in pairwise spatial relationships between cell types.</span></span>
<span><span class="va">spicyTest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/spicy.html" class="external-link">spicy</a></span><span class="op">(</span><span class="va">cells</span><span class="op">[</span>, <span class="va">cellsToUse</span><span class="op">]</span>, </span>
<span>                   condition <span class="op">=</span> <span class="st">"Status"</span>, </span>
<span>                   cellType <span class="op">=</span> <span class="st">"clusters"</span>,</span>
<span>                   imageID <span class="op">=</span> <span class="st">"imageID"</span>,</span>
<span>                   spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"m.cx"</span>, <span class="st">"m.cy"</span><span class="op">)</span>,</span>
<span>                   Rs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">50</span>, <span class="fl">100</span><span class="op">)</span>,</span>
<span>                   sigma <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                   BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/topPairs.html" class="external-link">topPairs</a></span><span class="op">(</span><span class="va">spicyTest</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                         intercept coefficient      p.value adj.pvalue</span></span>
<span><span class="co">## cluster_19__cluster_15 -150.88696   183.59375 0.0004249372  0.1534023</span></span>
<span><span class="co">## cluster_15__cluster_19 -152.01839   171.46981 0.0022585963  0.3537460</span></span>
<span><span class="co">## cluster_4__cluster_3    -86.85690    47.51106 0.0029397171  0.3537460</span></span>
<span><span class="co">## cluster_3__cluster_4    -81.83179    45.80377 0.0041898555  0.3781345</span></span>
<span><span class="co">## cluster_19__cluster_9  -113.22339    83.79345 0.0113279837  0.6195050</span></span>
<span><span class="co">## cluster_9__cluster_19  -115.90952    71.61334 0.0208610085  0.6195050</span></span>
<span><span class="co">## cluster_19__cluster_5  -127.85692    51.95523 0.0250966956  0.6195050</span></span>
<span><span class="co">## cluster_15__cluster_12 -143.45605    98.09055 0.0251142262  0.6195050</span></span>
<span><span class="co">## cluster_7__cluster_5     86.36437    66.75056 0.0263208116  0.6195050</span></span>
<span><span class="co">## cluster_5__cluster_19  -130.94590    47.54211 0.0276756032  0.6195050</span></span>
<span><span class="co">##                              from         to</span></span>
<span><span class="co">## cluster_19__cluster_15 cluster_19 cluster_15</span></span>
<span><span class="co">## cluster_15__cluster_19 cluster_15 cluster_19</span></span>
<span><span class="co">## cluster_4__cluster_3    cluster_4  cluster_3</span></span>
<span><span class="co">## cluster_3__cluster_4    cluster_3  cluster_4</span></span>
<span><span class="co">## cluster_19__cluster_9  cluster_19  cluster_9</span></span>
<span><span class="co">## cluster_9__cluster_19   cluster_9 cluster_19</span></span>
<span><span class="co">## cluster_19__cluster_5  cluster_19  cluster_5</span></span>
<span><span class="co">## cluster_15__cluster_12 cluster_15 cluster_12</span></span>
<span><span class="co">## cluster_7__cluster_5    cluster_7  cluster_5</span></span>
<span><span class="co">## cluster_5__cluster_19   cluster_5 cluster_19</span></span></code></pre>
<p>We can visualise these tests using <code>signifPlot</code> where we
observe that cell type pairs appear to become less attractive (or avoid
more) in the progression sample.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise which relationships are changing the most.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/signifPlot.html" class="external-link">signifPlot</a></span><span class="op">(</span><span class="va">spicyTest</span>,</span>
<span>           breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.5</span>, <span class="fl">3</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="lisaclust-find-cellular-neighbourhoods">lisaClust: Find cellular neighbourhoods<a class="anchor" aria-label="anchor" href="#lisaclust-find-cellular-neighbourhoods"></a>
</h2>
<p>Our lisaClust package (<a href="https://www.bioconductor.org/packages/devel/bioc/html/lisaClust.html" class="external-link uri">https://www.bioconductor.org/packages/devel/bioc/html/lisaClust.html</a>)[<a href="https://www.bioconductor.org/packages/devel/bioc/html/lisaClust.html" class="external-link uri">https://www.bioconductor.org/packages/devel/bioc/html/lisaClust.html</a>]
provides a series of functions to identify and visualise regions of
tissue where spatial associations between cell-types is similar. This
package can be used to provide a high-level summary of cell-type
colocalisation in multiplexed imaging data that has been segmented at a
single-cell resolution. Here we use the <code>lisaClust</code> function
to clusters cells into 5 regions with distinct spatial ordering.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Cluster cells into spatial regions with similar composition.</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/lisaClust.html" class="external-link">lisaClust</a></span><span class="op">(</span><span class="va">cells</span>, </span>
<span>                   k <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                   Rs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">50</span>, <span class="fl">100</span><span class="op">)</span>,</span>
<span>                   sigma <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                   spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"m.cx"</span>, <span class="st">"m.cy"</span><span class="op">)</span>, </span>
<span>                   cellType <span class="op">=</span> <span class="st">"clusters"</span>, </span>
<span>                   BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="region---cell-type-enrichment-heatmap">Region - cell type enrichment heatmap<a class="anchor" aria-label="anchor" href="#region---cell-type-enrichment-heatmap"></a>
</h3>
<p>We can try to interpret which spatial orderings the regions are
quantifying using the <code>regionMap</code> function. This plots the
frequency of each cell type in a region relative to what you would
expect by chance.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise the enrichment of each cell type in each region</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/regionMap.html" class="external-link">regionMap</a></span><span class="op">(</span><span class="va">cells</span>, cellType <span class="op">=</span> <span class="st">"clusters"</span>, limit <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-23-1.png" width="480"></p>
</div>
<div class="section level3">
<h3 id="visualise-regions">Visualise regions<a class="anchor" aria-label="anchor" href="#visualise-regions"></a>
</h3>
<p>By default, these identified regions are stored in the
<code>regions</code> column in the <code>colData</code> of our object.
We can quickly examine the spatial arrangement of these regions using
<code>ggplot</code>.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract cell information and filter to specific image.</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">imageID</span> <span class="op">==</span> <span class="st">"Point2206_pt1116_31620"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Colour cells by their region.</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">m.cx</span>, y <span class="op">=</span> <span class="va">m.cy</span>, colour <span class="op">=</span> <span class="va">region</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<p>While much slower, we have also implemented a function for overlaying
the region information as a hatching pattern so that the information can
be viewed simultaneously with the cell type calls.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use hatching to visualise regions and cell types.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/hatchingPlot.html" class="external-link">hatchingPlot</a></span><span class="op">(</span><span class="va">cells</span>,</span>
<span>             useImages <span class="op">=</span> <span class="st">"Point2206_pt1116_31620"</span>,</span>
<span>             cellType <span class="op">=</span> <span class="st">"clusters"</span>,</span>
<span>             spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"m.cx"</span>, <span class="st">"m.cy"</span><span class="op">)</span></span>
<span>            <span class="op">)</span></span></code></pre></div>
<p>This plot is a ggplot object and so the scale can be modified with
<code>scale_region_manual</code>.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use hatching to visualise regions and cell types.</span></span>
<span><span class="co"># Relabel the hatching of the regions.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/hatchingPlot.html" class="external-link">hatchingPlot</a></span><span class="op">(</span><span class="va">cells</span>,</span>
<span>             useImages <span class="op">=</span> <span class="st">"Point2206_pt1116_31620"</span>,</span>
<span>             cellType <span class="op">=</span> <span class="st">"clusters"</span>,</span>
<span>             spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"m.cx"</span>, <span class="st">"m.cy"</span><span class="op">)</span>,</span>
<span>             window <span class="op">=</span> <span class="st">"square"</span>,</span>
<span>             nbp <span class="op">=</span> <span class="fl">300</span>,</span>
<span>             line.spacing <span class="op">=</span> <span class="fl">41</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/scale_region.html" class="external-link">scale_region_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>region_1 <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                               region_2 <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                               region_3 <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                               region_4 <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                               region_5 <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  </span>
<span>   <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="test-for-association-with-progression">Test for association with progression<a class="anchor" aria-label="anchor" href="#test-for-association-with-progression"></a>
</h3>
<p>If needed, we can again quickly use the <code>colTest</code> function
to test for associations between the proportions of the cells in each
region and progression status using either Wilcoxon rank sum tests or
t-tests. Here we see a adjusted p-value less than 0.05.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test if the proportion of each region is associated</span></span>
<span><span class="co"># with progression status.</span></span>
<span><span class="va">testRegion</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/colTest.html" class="external-link">colTest</a></span><span class="op">(</span><span class="va">cells</span><span class="op">[</span>,<span class="va">cellsToUse</span><span class="op">]</span>, </span>
<span>                      feature <span class="op">=</span> <span class="st">"region"</span>,</span>
<span>                      condition <span class="op">=</span> <span class="st">"Status"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">testRegion</span></span></code></pre></div>
<pre><code><span><span class="co">##            W pval adjPval  cluster</span></span>
<span><span class="co">## region_4 380 0.22    0.77 region_4</span></span>
<span><span class="co">## region_2 260 0.44    0.77 region_2</span></span>
<span><span class="co">## region_5 270 0.46    0.77 region_5</span></span>
<span><span class="co">## region_3 330 0.74    0.88 region_3</span></span>
<span><span class="co">## region_1 320 0.88    0.88 region_1</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="classifyr-classification">ClassifyR: Classification<a class="anchor" aria-label="anchor" href="#classifyr-classification"></a>
</h2>
<p>Our ClassifyR package, <a href="https://github.com/SydneyBioX/ClassifyR" class="external-link">https://github.com/SydneyBioX/ClassifyR</a>,
formalises a convenient framework for evaluating classification in R. We
provide functionality to easily include four key modelling stages; Data
transformation, feature selection, classifier training and prediction;
into a cross-validation loop. Here we use the <code>crossValidate</code>
function to perform 100 repeats of 5-fold cross-validation to evaluate
the performance of an elastic net model applied to three quantification
of our MIBI-TOF data; cell type proportions, average mean of each cell
type and region proportions.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create list to store data.frames</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add proportions of each cell type in each image</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"props"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/getProp.html" class="external-link">getProp</a></span><span class="op">(</span><span class="va">cells</span>, <span class="st">"clusters"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add pairwise associations</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"dist"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/getPairwise.html" class="external-link">getPairwise</a></span><span class="op">(</span><span class="va">cells</span>, spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"m.cx"</span>, <span class="st">"m.cy"</span><span class="op">)</span>,</span>
<span>                              cellType <span class="op">=</span> <span class="st">"clusters"</span>,</span>
<span>                              Rs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">50</span>, <span class="fl">100</span><span class="op">)</span>,</span>
<span>                              sigma <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                              BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"dist"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">data</span><span class="op">[[</span><span class="st">"dist"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Add proportions of each region in each image</span></span>
<span><span class="co"># to the list of dataframes.</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"regions"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/getProp.html" class="external-link">getProp</a></span><span class="op">(</span><span class="va">cells</span>, <span class="st">"region"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset data images with progression status and NA clinical variables.</span></span>
<span><span class="va">measurements</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">data</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="va">x</span><span class="op">[</span><span class="va">imagesToUse</span>, <span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set seed</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform cross-validation of an elastic net model</span></span>
<span><span class="co"># with 100 repeats of 5-fold cross-validation.</span></span>
<span><span class="va">cv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ClassifyR/man/crossValidate.html" class="external-link">crossValidate</a></span><span class="op">(</span>measurements <span class="op">=</span> <span class="va">measurements</span>, </span>
<span>                    outcome <span class="op">=</span> <span class="va">clinical</span><span class="op">[</span><span class="va">imagesToUse</span>, <span class="st">"Status"</span><span class="op">]</span>,</span>
<span>                    classifier <span class="op">=</span> <span class="st">"GLM"</span>,</span>
<span>                    nFolds <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                    nRepeats <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                    nCores <span class="op">=</span> <span class="va">nCores</span></span>
<span>                    <span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="visualise-cross-validated-prediction-performance">Visualise cross-validated prediction performance<a class="anchor" aria-label="anchor" href="#visualise-cross-validated-prediction-performance"></a>
</h3>
<p>Here we use the <code>performancePlot</code> function to assess the
AUC from each repeat of the 5-fold cross-validation. We see that the
lisaClust regions appear to capture information which is predictive of
progression status of the patients.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate AUC for each cross-validation repeat and plot.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ClassifyR/man/performancePlot.html" class="external-link">performancePlot</a></span><span class="op">(</span><span class="va">cv</span>,</span>
<span>                performanceName <span class="op">=</span> <span class="st">"AUC"</span>,</span>
<span>                characteristicsList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Assay Name"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>Here we have used a pipeline of our spatial analysis R packages to
demonstrate an easy way to segment, cluster, normalise, quantify and
classify high dimensional in situ cytometry data all within R.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Alexander Nicholls, Nicholas Robertson, Nicolas Canete, Elijah Willie, Ellis Patrick.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
