<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="spicyWorkflow">
<title>Performing a spatial analysis of multiplexed tissue imaging data. • spicyWorkflow</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Performing a spatial analysis of multiplexed tissue imaging data.">
<meta property="og:description" content="spicyWorkflow">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">spicyWorkflow</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/spicyWorkflow.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/SydneyBioX/spicyWorkflow/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="spicyWorkflow_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Performing a spatial analysis of multiplexed tissue imaging data.</h1>
                        <h4 data-toc-skip class="author">Alexander Nicholls</h4>
            <address class="author_afil">
      Westmead Institute for Medical Research, University of Sydney, Australia<br><h4 data-toc-skip class="author">Nicholas Robertson</h4>
            <address class="author_afil">
      School of Mathematics and Statistics, University of Sydney, Australia<br><h4 data-toc-skip class="author">Nicholas Canete</h4>
            <address class="author_afil">
      Westmead Institute for Medical Research, University of Sydney, Australia<br><h4 data-toc-skip class="author">Elijah Willie</h4>
            <address class="author_afil">
      Westmead Institute for Medical Research, University of Sydney, AustraliaSchool of Mathematics and Statistics, University of Sydney, Australia<br><h4 data-toc-skip class="author">Ellis Patrick</h4>
            <address class="author_afil">
      Westmead Institute for Medical Research, University of Sydney, AustraliaSchool of Mathematics and Statistics, University of Sydney, Australia<br><h4 data-toc-skip class="date">27 July, 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/SydneyBioX/spicyWorkflow/blob/HEAD/vignettes/spicyWorkflow.Rmd" class="external-link"><code>vignettes/spicyWorkflow.Rmd</code></a></small>
      <div class="d-none name"><code>spicyWorkflow.Rmd</code></div>
    </address>
</address>
</address>
</address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="version-info">Version Info<a class="anchor" aria-label="anchor" href="#version-info"></a>
</h2>
<p>
</p>
<p><strong>R version</strong>: R version 4.3.0 alpha (2023-03-29 r84123) <br><strong>Bioconductor version</strong>: 3.17 <br></p>
<p>Understanding the interplay between different types of cells and their immediate environment is critical for understanding the mechanisms of cells themselves and their function in the context of human diseases. Recent advances in high dimensional in situ cytometry technologies have fundamentally revolutionised our ability to observe these complex cellular relationships providing an unprecedented characterisation of cellular heterogeneity in a tissue environment.</p>
<p>We have developed an analytical framework for analysing data from high dimensional in situ cytometry assays including CODEX, CycIF, IMC and High Definition Spatial Transcriptomics. Implemented in R, this framework makes use of functionality from our Bioconductor packages spicyR, lisaClust, treekoR, FuseSOM, simpleSeg and ClassifyR. Below we will provide an overview of key steps which are needed to interrogate the comprehensive spatial information generated by these exciting new technologies including cell segmentation, feature normalisation, cell type identification, micro-environment characterisation, spatial hypothesis testing and patient classification. Ultimately, our modular analysis framework provides a cohesive and accessible entry point into spatially resolved single cell data analysis for any R-based bioinformatics.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install the current release of spicyWorkflow, run the following code.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"spicyWorkflow"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="loading-r-packages">Loading R packages<a class="anchor" aria-label="anchor" href="#loading-r-packages"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BodenmillerGroup/cytomapper" class="external-link">cytomapper</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/SydneyBioX/simpleSeg" class="external-link">simpleSeg</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">FuseSOM</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/" class="external-link">ggpubr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">spicyR</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/ClassifyR/" class="external-link">ClassifyR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">lisaClust</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="global-paramaters">Global paramaters<a class="anchor" aria-label="anchor" href="#global-paramaters"></a>
</h2>
<p>It is convenient to set the number of cores for running code in parallel. Please choose a number that is appropriate for your resources.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nCores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">BPPARAM</span> <span class="op">&lt;-</span> <span class="fu">simpleSeg</span><span class="fu">:::</span><span class="fu"><a href="https://rdrr.io/pkg/simpleSeg/man/generateBPParam.html" class="external-link">generateBPParam</a></span><span class="op">(</span><span class="va">nCores</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="context">Context<a class="anchor" aria-label="anchor" href="#context"></a>
</h2>
<p>In the following we will re-analyse some MIBI-TOF data <a href="https://www.sciencedirect.com/science/article/pii/S0092867421014860?via%3Dihub#!" class="external-link">(Risom et al, 2022)</a> profiling the spatial landscape of ductal carcinoma in situ (DCIS), which is a pre-invasive lesion that is thought to be a precursor to invasive breast cancer (IBC). The key conclusion of this manuscript (amongst others) is that spatial information about cells can be used to predict disease progression in patients. We will use our spicy workflow to make a similar conclusion.</p>
<p>The R code for this analysis is available on github <a href="https://github.com/SydneyBioX/spicyWorkflow" class="external-link">https://github.com/SydneyBioX/spicyWorkflow</a>. A mildly <a href="https://github.com/SydneyBioX/spicyWorkflow/blob/master/organisePublishedData.R" class="external-link">processed</a> version of the data used in the manuscript is available in this repository.</p>
</div>
<div class="section level2">
<h2 id="read-in-images">Read in images<a class="anchor" aria-label="anchor" href="#read-in-images"></a>
</h2>
<p>The images are stored in the <code>images</code> folder within the <code>data</code> folder. Here we use <code><a href="https://rdrr.io/pkg/cytomapper/man/loadImages.html" class="external-link">loadImages()</a></code> from the <code>cytomapper</code> package to load all the tiff images into a <code>CytoImageList</code> object and store the images as h5 file on-disk.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pathToImages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/images"</span>, package <span class="op">=</span> <span class="st">"spicyWorkflow"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Store images in a CytoImageList with images on_disk as h5 files to save memory.</span></span>
<span><span class="va">images</span> <span class="op">&lt;-</span> <span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/loadImages.html" class="external-link">loadImages</a></span><span class="op">(</span></span>
<span>  <span class="va">pathToImages</span>,</span>
<span>  single_channel <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  on_disk <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  h5FilesPath <span class="op">=</span> <span class="fu">HDF5Array</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/HDF5Array/man/dump-management.html" class="external-link">getHDF5DumpDir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/gc.html" class="external-link">gc</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            used  (Mb) gc trigger   (Mb) max used  (Mb)</span></span>
<span><span class="co">## Ncells 11340591 605.7   19359011 1033.9 14908772 796.3</span></span>
<span><span class="co">## Vcells 18692386 142.7   33938162  259.0 23445664 178.9</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="load-the-clinical-data">Load the clinical data<a class="anchor" aria-label="anchor" href="#load-the-clinical-data"></a>
</h2>
<p>To associate features in our image with disease progression, it is important to read in information which links image identifiers to their progression status. We will do this here, making sure that our <code>imageID</code> match. ## Read the clinical data</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in clinical data, manipulate imageID and select columns</span></span>
<span><span class="va">clinical</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"extdata/1-s2.0-S0092867421014860-mmc1.csv"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"spicyWorkflow"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">clinical</span> <span class="op">&lt;-</span> <span class="va">clinical</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>imageID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Point"</span>, <span class="va">PointNumber</span>, <span class="st">"_pt"</span>, <span class="va">Patient_ID</span>, <span class="st">"_"</span>, <span class="va">TMAD_Patient</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">clinical</span><span class="op">$</span><span class="va">imageID</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"normal"</span>, <span class="va">clinical</span><span class="op">$</span><span class="va">Tissue_Type</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">$</span><span class="va">imageID</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"normal"</span>, <span class="va">clinical</span><span class="op">$</span><span class="va">Tissue_Type</span><span class="op">)</span><span class="op">]</span>, <span class="st">"_Normal"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">clinicalVariables</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"imageID"</span>, <span class="st">"Patient_ID"</span>, <span class="st">"Status"</span>, <span class="st">"Age"</span>, <span class="st">"SUBTYPE"</span>, <span class="st">"PAM50"</span>, <span class="st">"Treatment"</span>, <span class="st">"DCIS_grade"</span>, <span class="st">"Necrosis"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">clinical</span><span class="op">$</span><span class="va">imageID</span></span></code></pre></div>
<div class="section level3">
<h3 id="put-the-clinical-data-into-the-coldata-of-singlecellexperiment">Put the clinical data into the colData of SingleCellExperiment<a class="anchor" aria-label="anchor" href="#put-the-clinical-data-into-the-coldata-of-singlecellexperiment"></a>
</h3>
<p>We can then store the clinical information in the <code>mcols</code> of the <code>CytoImageList</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add the clinical data to mcols of images.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">clinical</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">images</span><span class="op">)</span>, <span class="va">clinicalVariables</span><span class="op">]</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="simpleseg-segment-the-cells-in-the-images">SimpleSeg: Segment the cells in the images<a class="anchor" aria-label="anchor" href="#simpleseg-segment-the-cells-in-the-images"></a>
</h2>
<p>Our simpleSeg R package on <a href="https://github.com/SydneyBioX/simpleSeg" class="external-link">https://github.com/SydneyBioX/simpleSeg</a> provides a series of functions to generate simple segmentation masks of images. These functions leverage the functionality of the <a href="https://bioconductor.org/packages/release/bioc/vignettes/EBImage/inst/doc/EBImage-introduction.html" class="external-link">EBImage</a> package on Bioconductor. For more flexibility when performing your segmentation in R we recommend learning to use the EBimage package. A key strength of the simpleSeg package is that we have coded multiple ways to perform some simple segmentation operations as well as incorporating multiple automatic procedures to optimise some key parameters when these aren’t specified.</p>
<div class="section level3">
<h3 id="run-simpleseg">Run simpleSeg<a class="anchor" aria-label="anchor" href="#run-simpleseg"></a>
</h3>
<p>If your images are stored in a <code>list</code> or <code>CytoImageList</code> they can be segmented with a simple call to <code><a href="https://rdrr.io/pkg/simpleSeg/man/simpleSeg.html" class="external-link">simpleSeg()</a></code>. Here we have ask <code>simpleSeg</code> to do multiple things. First, we would like to use a combination of principal component analysis of all channels guided by the H33 channel to summarise the nuclei signal in the images. Secondly, to estimate the cell body of the cells we will simply dilate out from the nuclei by 2 pixels. We have also requested that the channels be square root transformed and that a minimum cell size of 40 pixels be used as a size selection step.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate segmentation masks</span></span>
<span><span class="va">masks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/simpleSeg/man/simpleSeg.html" class="external-link">simpleSeg</a></span><span class="op">(</span></span>
<span>  <span class="va">images</span>,</span>
<span>  nucleus <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"HH3"</span><span class="op">)</span>,</span>
<span>  cellBody <span class="op">=</span> <span class="st">"dilate"</span>,</span>
<span>  transform <span class="op">=</span> <span class="st">"sqrt"</span>,</span>
<span>  sizeSelection <span class="op">=</span> <span class="fl">40</span>,</span>
<span>  discSize <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  pca <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  cores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualise-separation">Visualise separation<a class="anchor" aria-label="anchor" href="#visualise-separation"></a>
</h3>
<p>The <code>display</code> and <code>colorLabels</code> functions in <code>EBImage</code> make it very easy to examine the performance of the cell segmentation. The great thing about <code>display</code> is that if used in an interactive session it is very easy to zoom in and out of the image.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise segmentation performance one way.</span></span>
<span><span class="fu">EBImage</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/colorLabels.html" class="external-link">colorLabels</a></span><span class="op">(</span><span class="va">masks</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/visualise%20segmentation-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="visualise-outlines">Visualise outlines<a class="anchor" aria-label="anchor" href="#visualise-outlines"></a>
</h3>
<p>The <code>plotPixels</code> function in <code>cytomapper</code> make it easy to overlay the masks on top of the intensities of 6 markers. Here we can see that the segmentation appears to be performing reasonably.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise segmentation performance another way.</span></span>
<span><span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/plotPixels.html" class="external-link">plotPixels</a></span><span class="op">(</span></span>
<span>  image <span class="op">=</span> <span class="va">images</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>  mask <span class="op">=</span> <span class="va">masks</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>  img_id <span class="op">=</span> <span class="st">"imageID"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PanKRT"</span>, <span class="st">"GLUT1"</span>, <span class="st">"HH3"</span>, <span class="st">"CD3"</span>, <span class="st">"CD20"</span><span class="op">)</span>,</span>
<span>  display <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>  colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    HH3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"blue"</span><span class="op">)</span>,</span>
<span>    CD3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"purple"</span><span class="op">)</span>,</span>
<span>    CD20 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"green"</span><span class="op">)</span>,</span>
<span>    GLUT1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span>,</span>
<span>    PanKRT <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"yellow"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  bcg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    HH3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1.5</span><span class="op">)</span>,</span>
<span>    CD3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1.5</span><span class="op">)</span>,</span>
<span>    CD20 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1.5</span><span class="op">)</span>,</span>
<span>    GLUT1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1.5</span><span class="op">)</span>,</span>
<span>    PanKRT <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1.5</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="summarise-cell-features-">Summarise cell features.<a class="anchor" aria-label="anchor" href="#summarise-cell-features-"></a>
</h2>
<p>In order to characterise the phenotypes of each of the segmented cells, <code>measureObjects</code> from <code>cytomapper</code> will calculate the average intensity of each channel within each cell as well as a few morphological features. The channel intensities will be stored in the <code>counts assay</code> in a <code>SingleCellExperiment</code>. Information on the spatial location of each cell is stored in <code>colData</code> in the <code>m.cx</code> and <code>m.cy</code> columns. In addition to this, it will propagate the information we have store in the <code>mcols</code> of our <code>CytoImageList</code> in the <code>colData</code> of the resulting <code>SingleCellExperiment</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Summarise the expression of each marker in each cell</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu">cytomapper</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cytomapper/man/measureObjects.html" class="external-link">measureObjects</a></span><span class="op">(</span></span>
<span>  <span class="va">masks</span>,</span>
<span>  <span class="va">images</span>,</span>
<span>  img_id <span class="op">=</span> <span class="st">"imageID"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="normalise-data">Normalise data<a class="anchor" aria-label="anchor" href="#normalise-data"></a>
</h2>
<p>We should check to see if the marker intensities of each cell require some form of transformation or normalisation. Here we extract the intensities from the <code>counts</code> assay. Looking at CK7 which should be expressed in the majority of the tumour cells, the intensities are clearly very skewed.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract marker data and bind with information about images</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">cells</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plots densities of CK7 for each image.</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">CK7</span>, colour <span class="op">=</span> <span class="va">imageID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-5-1.png" width="480"></p>
<p>We can transform and normalise our data using the <code>normalizeCells</code> function. Here we have taken the intensities from the <code>counts</code> assay, performed a square root transform, then for each image trimmed the 99 quantile and min-max scaled to 0-1. This modified data is then stored in the <code>norm</code> assay by default. We can see that this normalised data appears more bimodal, not perfectly, but likely to a sufficient degree for clustering.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Transform and normalise the marker expression of each cell type.</span></span>
<span><span class="co"># Use a square root transform, then trimmed the 99 quantile</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/simpleSeg/man/normalizeCells.html" class="external-link">normalizeCells</a></span><span class="op">(</span><span class="va">cells</span>,</span>
<span>  transformation <span class="op">=</span> <span class="st">"asinh"</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"trim99"</span>, <span class="st">"minMax"</span>, <span class="st">"PC1"</span><span class="op">)</span>,</span>
<span>  assayIn <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  cores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract normalised marker information.</span></span>
<span><span class="va">norm_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">cells</span>, <span class="st">"norm"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Plots densities of normalised CK7 for each image.</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">norm_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">CK7</span>, colour <span class="op">=</span> <span class="va">imageID</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-6-1.png" width="480"></p>
</div>
<div class="section level2">
<h2 id="fusesom-cluster-cells-into-cell-types">FuseSOM: Cluster cells into cell types<a class="anchor" aria-label="anchor" href="#fusesom-cluster-cells-into-cell-types"></a>
</h2>
<p>Our FuseSOM R package on <a href="https://github.com/ecool50/FuseSOM" class="external-link">https://github.com/ecool50/FuseSOM</a> and provides a pipeline for the clustering of highly multiplexed in situ imaging cytometry assays. This pipeline uses the Self Organising Map architecture coupled with Multiview hierarchical clustering and provides functions for the estimation of the number of clusters.</p>
<p>Here we cluster using the <code>runFuseSOM</code> function. We have chosen to specify the same subset of markers used in the original manuscript for gating cell types. We have also specified the number of clusters to identify to be <code>numClusters = 24</code>. In addition to this, while FuseSOM can automatically estimate a grid size for the self organising map.</p>
<div class="section level3">
<h3 id="perform-the-clustering">Perform the clustering<a class="anchor" aria-label="anchor" href="#perform-the-clustering"></a>
</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The markers used in the original publication to gate cell types.</span></span>
<span><span class="va">useMarkers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"PanKRT"</span>, <span class="st">"ECAD"</span>, <span class="st">"CK7"</span>, <span class="st">"VIM"</span>, <span class="st">"FAP"</span>, <span class="st">"CD31"</span>, <span class="st">"CK5"</span>, <span class="st">"SMA"</span>,</span>
<span>  <span class="st">"CD45"</span>, <span class="st">"CD4"</span>, <span class="st">"CD3"</span>, <span class="st">"CD8"</span>, <span class="st">"CD20"</span>, <span class="st">"CD68"</span>, <span class="st">"CD14"</span>, <span class="st">"CD11c"</span>,</span>
<span>  <span class="st">"HLADRDPDQ"</span>, <span class="st">"MPO"</span>, <span class="st">"Tryptase"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set seed.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate SOM and cluster cells into 20 groups.</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/runFuseSOM.html" class="external-link">runFuseSOM</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  markers <span class="op">=</span> <span class="va">useMarkers</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"norm"</span>,</span>
<span>  numClusters <span class="op">=</span> <span class="fl">24</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="attempt-to-interpret-the-phenotype-of-each-cluster">Attempt to interpret the phenotype of each cluster<a class="anchor" aria-label="anchor" href="#attempt-to-interpret-the-phenotype-of-each-cluster"></a>
</h3>
<p>We can begin the process of understanding what each of these cell clusters are by using the <code>plotGroupedHeatmap</code> function from <code>scater</code>. At the least, here we can see we capture all the major immune populations that we expect to see.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise marker expression in each cluster.</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotGroupedHeatmap.html" class="external-link">plotGroupedHeatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  features <span class="op">=</span> <span class="va">useMarkers</span>,</span>
<span>  group <span class="op">=</span> <span class="st">"clusters"</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"norm"</span>,</span>
<span>  center <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  zlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  cluster_rows <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="check-how-many-clusters-should-be-used-">Check how many clusters should be used.<a class="anchor" aria-label="anchor" href="#check-how-many-clusters-should-be-used-"></a>
</h3>
<p>We can check to see how reasonable our choice of 24 clusters is using the <code>estimateNumCluster</code> and the <code>optiPlot</code> functions. Here we examine the Gap method, others such as Silhouette and Within Cluster Distance are also available.</p>
<p>As we can be seen below, we chose the second elbow point as the optimal number of clusters.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate metrics for estimating the number of clusters.</span></span>
<span><span class="co"># As I've already run runFuseSOM I don't need to run generateSOM().</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/estimateNumCluster.html" class="external-link">estimateNumCluster</a></span><span class="op">(</span><span class="va">cells</span>, kSeq <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/FuseSOM/man/optiPlot.html" class="external-link">optiPlot</a></span><span class="op">(</span><span class="va">cells</span>, method <span class="op">=</span> <span class="st">"gap"</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="check-cluster-frequencies">Check cluster frequencies<a class="anchor" aria-label="anchor" href="#check-cluster-frequencies"></a>
</h3>
<p>We find it always useful to check the number of cells in each cluster. Here we can see that cluster 4 is contains lots of (most likely tumour) cells and cluster 16 contains very few cells.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check cluster frequencies.</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span><span class="op">$</span><span class="va">clusters</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## cluster_17  cluster_9 cluster_21  cluster_7 cluster_12  cluster_3  cluster_2 </span></span>
<span><span class="co">##        205        254        358        593        664        695        711 </span></span>
<span><span class="co">## cluster_23 cluster_14  cluster_8 cluster_18 cluster_22 cluster_19 cluster_13 </span></span>
<span><span class="co">##        719        840        890       1019       1154       1236       1354 </span></span>
<span><span class="co">##  cluster_1 cluster_10 cluster_16 cluster_11 cluster_20 cluster_24 cluster_15 </span></span>
<span><span class="co">##       1382       1431       1614       1871       3481       3965       4403 </span></span>
<span><span class="co">##  cluster_5  cluster_4  cluster_6 </span></span>
<span><span class="co">##       6166       6551      19778</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="dimension-reduction">Dimension reduction<a class="anchor" aria-label="anchor" href="#dimension-reduction"></a>
</h3>
<p>As our data is stored in a <code>SingleCellExperiment</code> we can also use <code>scater</code> to perform and visualise our data in a lower dimension to look for cluster differences.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span><span class="co"># Perform dimension reduction using UMP.</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  subset_row <span class="op">=</span> <span class="va">useMarkers</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"norm"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select a subset of images to plot.</span></span>
<span><span class="va">someImages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span><span class="op">$</span><span class="va">imageID</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">10</span>, <span class="fl">20</span>, <span class="fl">40</span>, <span class="fl">50</span>, <span class="fl">60</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span></span>
<span><span class="co"># UMAP by cell type cluster.</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span><span class="op">[</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span><span class="op">$</span><span class="va">imageID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">someImages</span><span class="op">]</span>,</span>
<span>  dimred <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"clusters"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="test-for-association-between-the-proportion-of-each-cell-type-and-progression-status">Test For association between the proportion of each cell type and progression status<a class="anchor" aria-label="anchor" href="#test-for-association-between-the-proportion-of-each-cell-type-and-progression-status"></a>
</h2>
<p>We recommend using a package such as <code>diffcyt</code> for testing for changes in abundance of cell types. However, the <code>colTest</code> function allows us to quickly test for associations between the proportions of the cell types and progression status using either Wilcoxon rank sum tests or t-tests. Here we see a p-value less than 0.05, but this does not equate to a small FDR.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Select cells which belong to individuals with progressor status.</span></span>
<span><span class="va">cellsToUse</span> <span class="op">&lt;-</span> <span class="va">cells</span><span class="op">$</span><span class="va">Status</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nonprogressor"</span>, <span class="st">"progressor"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform simple wicoxon rank sum tests on the columns of the proportion matrix.</span></span>
<span><span class="va">testProp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/colTest.html" class="external-link">colTest</a></span><span class="op">(</span><span class="va">cells</span><span class="op">[</span>, <span class="va">cellsToUse</span><span class="op">]</span>,</span>
<span>  condition <span class="op">=</span> <span class="st">"Status"</span>,</span>
<span>  feature <span class="op">=</span> <span class="st">"clusters"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">testProp</span></span></code></pre></div>
<pre><code><span><span class="co">##            mean in group nonprogressor mean in group progressor      t  pval</span></span>
<span><span class="co">## cluster_18                      0.0210                   0.0120  2.500 0.016</span></span>
<span><span class="co">## cluster_15                      0.0780                   0.0590  1.900 0.063</span></span>
<span><span class="co">## cluster_16                      0.0270                   0.0180  1.900 0.070</span></span>
<span><span class="co">## cluster_3                       0.0130                   0.0034  1.800 0.074</span></span>
<span><span class="co">## cluster_19                      0.0230                   0.0190  1.500 0.150</span></span>
<span><span class="co">## cluster_17                      0.0029                   0.0016  1.400 0.170</span></span>
<span><span class="co">## cluster_4                       0.0980                   0.1400 -1.400 0.170</span></span>
<span><span class="co">## cluster_24                      0.0590                   0.0390  1.300 0.190</span></span>
<span><span class="co">## cluster_9                       0.0046                   0.0027  1.300 0.210</span></span>
<span><span class="co">## cluster_5                       0.0960                   0.1300 -1.100 0.270</span></span>
<span><span class="co">## cluster_13                      0.0220                   0.0190  1.100 0.290</span></span>
<span><span class="co">## cluster_10                      0.0210                   0.0280 -0.820 0.420</span></span>
<span><span class="co">## cluster_11                      0.0200                   0.0250 -0.790 0.440</span></span>
<span><span class="co">## cluster_6                       0.3400                   0.3200  0.770 0.450</span></span>
<span><span class="co">## cluster_2                       0.0100                   0.0120 -0.740 0.470</span></span>
<span><span class="co">## cluster_14                      0.0120                   0.0130 -0.520 0.610</span></span>
<span><span class="co">## cluster_12                      0.0110                   0.0120 -0.510 0.620</span></span>
<span><span class="co">## cluster_22                      0.0180                   0.0210 -0.420 0.680</span></span>
<span><span class="co">## cluster_23                      0.0130                   0.0140 -0.420 0.680</span></span>
<span><span class="co">## cluster_7                       0.0093                   0.0130 -0.400 0.690</span></span>
<span><span class="co">## cluster_20                      0.0540                   0.0520  0.280 0.780</span></span>
<span><span class="co">## cluster_8                       0.0140                   0.0150 -0.190 0.850</span></span>
<span><span class="co">## cluster_1                       0.0240                   0.0230  0.140 0.890</span></span>
<span><span class="co">## cluster_21                      0.0066                   0.0065  0.055 0.960</span></span>
<span><span class="co">##            adjPval    cluster</span></span>
<span><span class="co">## cluster_18    0.38 cluster_18</span></span>
<span><span class="co">## cluster_15    0.44 cluster_15</span></span>
<span><span class="co">## cluster_16    0.44 cluster_16</span></span>
<span><span class="co">## cluster_3     0.44  cluster_3</span></span>
<span><span class="co">## cluster_19    0.56 cluster_19</span></span>
<span><span class="co">## cluster_17    0.56 cluster_17</span></span>
<span><span class="co">## cluster_4     0.56  cluster_4</span></span>
<span><span class="co">## cluster_24    0.56 cluster_24</span></span>
<span><span class="co">## cluster_9     0.56  cluster_9</span></span>
<span><span class="co">## cluster_5     0.63  cluster_5</span></span>
<span><span class="co">## cluster_13    0.63 cluster_13</span></span>
<span><span class="co">## cluster_10    0.75 cluster_10</span></span>
<span><span class="co">## cluster_11    0.75 cluster_11</span></span>
<span><span class="co">## cluster_6     0.75  cluster_6</span></span>
<span><span class="co">## cluster_2     0.75  cluster_2</span></span>
<span><span class="co">## cluster_14    0.83 cluster_14</span></span>
<span><span class="co">## cluster_12    0.83 cluster_12</span></span>
<span><span class="co">## cluster_22    0.83 cluster_22</span></span>
<span><span class="co">## cluster_23    0.83 cluster_23</span></span>
<span><span class="co">## cluster_7     0.83  cluster_7</span></span>
<span><span class="co">## cluster_20    0.89 cluster_20</span></span>
<span><span class="co">## cluster_8     0.93  cluster_8</span></span>
<span><span class="co">## cluster_1     0.93  cluster_1</span></span>
<span><span class="co">## cluster_21    0.96 cluster_21</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">imagesToUse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">clinical</span><span class="op">)</span><span class="op">[</span><span class="va">clinical</span><span class="op">[</span>, <span class="st">"Status"</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nonprogressor"</span>, <span class="st">"progressor"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">prop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/getProp.html" class="external-link">getProp</a></span><span class="op">(</span><span class="va">cells</span>, feature <span class="op">=</span> <span class="st">"clusters"</span><span class="op">)</span></span>
<span><span class="va">clusterToUse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">testProp</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="va">prop</span><span class="op">[</span><span class="va">imagesToUse</span>, <span class="va">clusterToUse</span><span class="op">]</span> <span class="op">~</span> <span class="va">clinical</span><span class="op">[</span><span class="va">imagesToUse</span>, <span class="st">"Status"</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="spicyr-test-spatial-relationships">spicyR: test spatial relationships<a class="anchor" aria-label="anchor" href="#spicyr-test-spatial-relationships"></a>
</h2>
<p>Our spicyR package (<a href="https://www.bioconductor.org/packages/devel/bioc/html/spicyR.html" class="external-link uri">https://www.bioconductor.org/packages/devel/bioc/html/spicyR.html</a>)[<a href="https://www.bioconductor.org/packages/devel/bioc/html/spicyR.html" class="external-link uri">https://www.bioconductor.org/packages/devel/bioc/html/spicyR.html</a>] provides a series of functions to aid in the analysis of both immunofluorescence and mass cytometry imaging data as well as other assays that can deeply phenotype individual cells and their spatial location. Here we use the <code>spicy</code> function to test for changes in the spatial relationships between pair-wise combinations of cells. We quantify spatial relationships using a combination of three radii <code>Rs = c(20, 50, 100)</code> and mildly account for some global tissue structure using <code>sigma = 50</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test for changes in pair-wise spatial relationships between cell types.</span></span>
<span><span class="va">spicyTest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/spicy.html" class="external-link">spicy</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span><span class="op">[</span>, <span class="va">cellsToUse</span><span class="op">]</span>,</span>
<span>  condition <span class="op">=</span> <span class="st">"Status"</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"clusters"</span>,</span>
<span>  imageID <span class="op">=</span> <span class="st">"imageID"</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"m.cx"</span>, <span class="st">"m.cy"</span><span class="op">)</span>,</span>
<span>  Rs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">50</span>, <span class="fl">100</span><span class="op">)</span>,</span>
<span>  sigma <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/topPairs.html" class="external-link">topPairs</a></span><span class="op">(</span><span class="va">spicyTest</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                          intercept coefficient     p.value adj.pvalue</span></span>
<span><span class="co">## cluster_9__cluster_21  -114.962528   243.01532 0.001061505  0.3131976</span></span>
<span><span class="co">## cluster_21__cluster_9  -114.039961   238.67211 0.001519455  0.3131976</span></span>
<span><span class="co">## cluster_10__cluster_11  -97.300723    91.72319 0.001739253  0.3131976</span></span>
<span><span class="co">## cluster_20__cluster_15    1.381275    44.36284 0.002661354  0.3131976</span></span>
<span><span class="co">## cluster_15__cluster_23   33.123975    80.55381 0.002718729  0.3131976</span></span>
<span><span class="co">## cluster_11__cluster_10  -91.995699    82.85717 0.004711349  0.3514033</span></span>
<span><span class="co">## cluster_11__cluster_11  123.810238   173.86550 0.005250569  0.3514033</span></span>
<span><span class="co">## cluster_23__cluster_15   36.773523    80.07047 0.005467489  0.3514033</span></span>
<span><span class="co">## cluster_15__cluster_20   -2.489198    39.95061 0.005490676  0.3514033</span></span>
<span><span class="co">## cluster_11__cluster_22   -2.818209   -99.24727 0.007230443  0.4085725</span></span>
<span><span class="co">##                              from         to</span></span>
<span><span class="co">## cluster_9__cluster_21   cluster_9 cluster_21</span></span>
<span><span class="co">## cluster_21__cluster_9  cluster_21  cluster_9</span></span>
<span><span class="co">## cluster_10__cluster_11 cluster_10 cluster_11</span></span>
<span><span class="co">## cluster_20__cluster_15 cluster_20 cluster_15</span></span>
<span><span class="co">## cluster_15__cluster_23 cluster_15 cluster_23</span></span>
<span><span class="co">## cluster_11__cluster_10 cluster_11 cluster_10</span></span>
<span><span class="co">## cluster_11__cluster_11 cluster_11 cluster_11</span></span>
<span><span class="co">## cluster_23__cluster_15 cluster_23 cluster_15</span></span>
<span><span class="co">## cluster_15__cluster_20 cluster_15 cluster_20</span></span>
<span><span class="co">## cluster_11__cluster_22 cluster_11 cluster_22</span></span></code></pre>
<p>We can visualise these tests using <code>signifPlot</code> where we observe that cell type pairs appear to become less attractive (or avoid more) in the progression sample.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise which relationships are changing the most.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/signifPlot.html" class="external-link">signifPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">spicyTest</span>,</span>
<span>  breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.5</span>, <span class="fl">3</span>, <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="lisaclust-find-cellular-neighbourhoods">lisaClust: Find cellular neighbourhoods<a class="anchor" aria-label="anchor" href="#lisaclust-find-cellular-neighbourhoods"></a>
</h2>
<p>Our lisaClust package (<a href="https://www.bioconductor.org/packages/devel/bioc/html/lisaClust.html" class="external-link uri">https://www.bioconductor.org/packages/devel/bioc/html/lisaClust.html</a>)[<a href="https://www.bioconductor.org/packages/devel/bioc/html/lisaClust.html" class="external-link uri">https://www.bioconductor.org/packages/devel/bioc/html/lisaClust.html</a>] provides a series of functions to identify and visualise regions of tissue where spatial associations between cell-types is similar. This package can be used to provide a high-level summary of cell-type co-localisation in multiplexed imaging data that has been segmented at a single-cell resolution. Here we use the <code>lisaClust</code> function to clusters cells into 5 regions with distinct spatial ordering.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Cluster cells into spatial regions with similar composition.</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/lisaClust.html" class="external-link">lisaClust</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  Rs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">50</span>, <span class="fl">100</span><span class="op">)</span>,</span>
<span>  sigma <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"m.cx"</span>, <span class="st">"m.cy"</span><span class="op">)</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"clusters"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="region---cell-type-enrichment-heatmap">Region - cell type enrichment heatmap<a class="anchor" aria-label="anchor" href="#region---cell-type-enrichment-heatmap"></a>
</h3>
<p>We can try to interpret which spatial orderings the regions are quantifying using the <code>regionMap</code> function. This plots the frequency of each cell type in a region relative to what you would expect by chance.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualise the enrichment of each cell type in each region</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/regionMap.html" class="external-link">regionMap</a></span><span class="op">(</span><span class="va">cells</span>, cellType <span class="op">=</span> <span class="st">"clusters"</span>, limit <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-16-1.png" width="480"></p>
</div>
<div class="section level3">
<h3 id="visualise-regions">Visualise regions<a class="anchor" aria-label="anchor" href="#visualise-regions"></a>
</h3>
<p>By default, these identified regions are stored in the <code>regions</code> column in the <code>colData</code> of our object. We can quickly examine the spatial arrangement of these regions using <code>ggplot</code>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract cell information and filter to specific image.</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">imageID</span> <span class="op">==</span> <span class="st">"Point2206_pt1116_31620"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Colour cells by their region.</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">m.cx</span>, y <span class="op">=</span> <span class="va">m.cy</span>, colour <span class="op">=</span> <span class="va">region</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<p>While much slower, we have also implemented a function for overlaying the region information as a hatching pattern so that the information can be viewed simultaneously with the cell type calls.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use hatching to visualise regions and cell types.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/hatchingPlot.html" class="external-link">hatchingPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  useImages <span class="op">=</span> <span class="st">"Point2206_pt1116_31620"</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"clusters"</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"m.cx"</span>, <span class="st">"m.cy"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>This plot is a ggplot object and so the scale can be modified with <code>scale_region_manual</code>.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use hatching to visualise regions and cell types.</span></span>
<span><span class="co"># Relabel the hatching of the regions.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/hatchingPlot.html" class="external-link">hatchingPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  useImages <span class="op">=</span> <span class="st">"Point2206_pt1116_31620"</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"clusters"</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"m.cx"</span>, <span class="st">"m.cy"</span><span class="op">)</span>,</span>
<span>  window <span class="op">=</span> <span class="st">"square"</span>,</span>
<span>  nbp <span class="op">=</span> <span class="fl">300</span>,</span>
<span>  line.spacing <span class="op">=</span> <span class="fl">41</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/lisaClust/man/scale_region.html" class="external-link">scale_region_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    region_1 <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    region_2 <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    region_3 <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    region_4 <span class="op">=</span> <span class="fl">4</span>,</span>
<span>    region_5 <span class="op">=</span> <span class="fl">3</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="test-for-association-with-progression">Test for association with progression<a class="anchor" aria-label="anchor" href="#test-for-association-with-progression"></a>
</h3>
<p>If needed, we can again quickly use the <code>colTest</code> function to test for associations between the proportions of the cells in each region and progression status using either Wilcoxon rank sum tests or t-tests. Here we see an adjusted p-value less than 0.05.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test if the proportion of each region is associated</span></span>
<span><span class="co"># with progression status.</span></span>
<span><span class="va">testRegion</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/colTest.html" class="external-link">colTest</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span><span class="op">[</span>, <span class="va">cellsToUse</span><span class="op">]</span>,</span>
<span>  feature <span class="op">=</span> <span class="st">"region"</span>,</span>
<span>  condition <span class="op">=</span> <span class="st">"Status"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">testRegion</span></span></code></pre></div>
<pre><code><span><span class="co">##          mean in group nonprogressor mean in group progressor     t pval</span></span>
<span><span class="co">## region_3                       0.330                    0.290  1.60 0.11</span></span>
<span><span class="co">## region_4                       0.290                    0.250  1.70 0.11</span></span>
<span><span class="co">## region_5                       0.120                    0.170 -1.70 0.11</span></span>
<span><span class="co">## region_2                       0.180                    0.220 -1.50 0.15</span></span>
<span><span class="co">## region_1                       0.082                    0.081  0.03 0.98</span></span>
<span><span class="co">##          adjPval  cluster</span></span>
<span><span class="co">## region_3    0.18 region_3</span></span>
<span><span class="co">## region_4    0.18 region_4</span></span>
<span><span class="co">## region_5    0.18 region_5</span></span>
<span><span class="co">## region_2    0.19 region_2</span></span>
<span><span class="co">## region_1    0.98 region_1</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="classifyr-classification">ClassifyR: Classification<a class="anchor" aria-label="anchor" href="#classifyr-classification"></a>
</h2>
<p>Our ClassifyR package, <a href="https://github.com/SydneyBioX/ClassifyR" class="external-link">https://github.com/SydneyBioX/ClassifyR</a>, formalises a convenient framework for evaluating classification in R. We provide functionality to easily include four key modelling stages; Data transformation, feature selection, classifier training and prediction; into a cross-validation loop. Here we use the <code>crossValidate</code> function to perform 100 repeats of 5-fold cross-validation to evaluate the performance of an elastic net model applied to three quantification of our MIBI-TOF data; cell type proportions, average mean of each cell type and region proportions.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create list to store data.frames</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add proportions of each cell type in each image</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"props"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/getProp.html" class="external-link">getProp</a></span><span class="op">(</span><span class="va">cells</span>, <span class="st">"clusters"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add pair-wise associations</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"dist"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/getPairwise.html" class="external-link">getPairwise</a></span><span class="op">(</span></span>
<span>  <span class="va">cells</span>,</span>
<span>  spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"m.cx"</span>, <span class="st">"m.cy"</span><span class="op">)</span>,</span>
<span>  cellType <span class="op">=</span> <span class="st">"clusters"</span>,</span>
<span>  Rs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">50</span>, <span class="fl">100</span><span class="op">)</span>,</span>
<span>  sigma <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span></span>
<span><span class="op">)</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"dist"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">data</span><span class="op">[[</span><span class="st">"dist"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Add proportions of each region in each image</span></span>
<span><span class="co"># to the list of dataframes.</span></span>
<span><span class="va">data</span><span class="op">[[</span><span class="st">"regions"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/spicyR/man/getProp.html" class="external-link">getProp</a></span><span class="op">(</span><span class="va">cells</span>, <span class="st">"region"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset data images with progression status and NA clinical variables.</span></span>
<span><span class="va">measurements</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">data</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">[</span><span class="va">imagesToUse</span>, <span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set seed</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">51773</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform cross-validation of an elastic net model</span></span>
<span><span class="co"># with 100 repeats of 5-fold cross-validation.</span></span>
<span><span class="va">cv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/crossValidate.html" class="external-link">crossValidate</a></span><span class="op">(</span></span>
<span>  measurements <span class="op">=</span> <span class="va">measurements</span>,</span>
<span>  outcome <span class="op">=</span> <span class="va">clinical</span><span class="op">[</span><span class="va">imagesToUse</span>, <span class="st">"Status"</span><span class="op">]</span>,</span>
<span>  classifier <span class="op">=</span> <span class="st">"GLM"</span>,</span>
<span>  nFolds <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  nRepeats <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  nCores <span class="op">=</span> <span class="va">nCores</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="visualise-cross-validated-prediction-performance">Visualise cross-validated prediction performance<a class="anchor" aria-label="anchor" href="#visualise-cross-validated-prediction-performance"></a>
</h3>
<p>Here we use the <code>performancePlot</code> function to assess the AUC from each repeat of the 5-fold cross-validation. We see that the lisaClust regions appear to capture information which is predictive of progression status of the patients.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate AUC for each cross-validation repeat and plot.</span></span>
<span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/performancePlot.html" class="external-link">performancePlot</a></span><span class="op">(</span></span>
<span>  <span class="va">cv</span>,</span>
<span>  metric <span class="op">=</span> <span class="st">"AUC"</span>,</span>
<span>  characteristicsList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Assay Name"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="spicyWorkflow_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>Here we have used a pipeline of our spatial analysis R packages to demonstrate an easy way to segment, cluster, normalise, quantify and classify high dimensional in situ cytometry data all within R.</p>
</div>
<div class="section level2">
<h2 id="sessioninfo">sessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.0 alpha (2023-03-29 r84123)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Debian GNU/Linux 11 (bullseye)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /dora/nobackup/biostat/software/R-devel/lib/libRblas.so </span></span>
<span><span class="co">## LAPACK: /dora/nobackup/biostat/software/R-devel/lib/libRlapack.so;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Australia/Sydney</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] lisaClust_1.7.3             ClassifyR_3.3.25           </span></span>
<span><span class="co">##  [3] survival_3.5-5              BiocParallel_1.33.12       </span></span>
<span><span class="co">##  [5] MultiAssayExperiment_1.25.9 generics_0.1.3             </span></span>
<span><span class="co">##  [7] spicyR_1.11.6               scater_1.27.8              </span></span>
<span><span class="co">##  [9] scuttle_1.9.4               ggpubr_0.6.0               </span></span>
<span><span class="co">## [11] FuseSOM_1.1.3               simpleSeg_1.1.1            </span></span>
<span><span class="co">## [13] ggplot2_3.4.2               dplyr_1.1.1                </span></span>
<span><span class="co">## [15] cytomapper_1.11.2           SingleCellExperiment_1.21.1</span></span>
<span><span class="co">## [17] SummarizedExperiment_1.29.1 Biobase_2.59.0             </span></span>
<span><span class="co">## [19] GenomicRanges_1.51.4        GenomeInfoDb_1.35.16       </span></span>
<span><span class="co">## [21] IRanges_2.33.1              S4Vectors_0.37.5           </span></span>
<span><span class="co">## [23] BiocGenerics_0.45.3         MatrixGenerics_1.11.1      </span></span>
<span><span class="co">## [25] matrixStats_0.63.0          EBImage_4.41.2             </span></span>
<span><span class="co">## [27] BiocStyle_2.27.1           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] fs_1.6.1                  spatstat.sparse_3.0-1    </span></span>
<span><span class="co">##   [3] bitops_1.0-7              httr_1.4.5               </span></span>
<span><span class="co">##   [5] RColorBrewer_1.1-3        prabclus_2.3-2           </span></span>
<span><span class="co">##   [7] DataVisualizations_1.2.3  numDeriv_2016.8-1.1      </span></span>
<span><span class="co">##   [9] tools_4.3.0               backports_1.4.1          </span></span>
<span><span class="co">##  [11] utf8_1.2.3                R6_2.5.1                 </span></span>
<span><span class="co">##  [13] vegan_2.6-4               HDF5Array_1.27.0         </span></span>
<span><span class="co">##  [15] uwot_0.1.14               mgcv_1.8-42              </span></span>
<span><span class="co">##  [17] rhdf5filters_1.11.2       permute_0.9-7            </span></span>
<span><span class="co">##  [19] withr_2.5.0               sp_1.6-0                 </span></span>
<span><span class="co">##  [21] analogue_0.17-6           gridExtra_2.3            </span></span>
<span><span class="co">##  [23] cli_3.6.1                 textshaping_0.3.6        </span></span>
<span><span class="co">##  [25] spatstat.explore_3.1-0    profileModel_0.6.1       </span></span>
<span><span class="co">##  [27] labeling_0.4.2            sass_0.4.5               </span></span>
<span><span class="co">##  [29] diptest_0.76-0            robustbase_0.95-1        </span></span>
<span><span class="co">##  [31] brglm_0.7.2               nnls_1.4                 </span></span>
<span><span class="co">##  [33] spatstat.data_3.0-1       genefilter_1.81.3        </span></span>
<span><span class="co">##  [35] proxy_0.4-27              pkgdown_2.0.7            </span></span>
<span><span class="co">##  [37] systemfonts_1.0.4         yulab.utils_0.0.6        </span></span>
<span><span class="co">##  [39] svglite_2.1.1             R.utils_2.12.2           </span></span>
<span><span class="co">##  [41] limma_3.55.7              RSQLite_2.3.1            </span></span>
<span><span class="co">##  [43] gridGraphics_0.5-1        spatstat.random_3.1-4    </span></span>
<span><span class="co">##  [45] car_3.1-2                 scam_1.2-13              </span></span>
<span><span class="co">##  [47] Matrix_1.5-4              ggbeeswarm_0.7.1         </span></span>
<span><span class="co">##  [49] fansi_1.0.4               abind_1.4-5              </span></span>
<span><span class="co">##  [51] R.methodsS3_1.8.2         terra_1.7-23             </span></span>
<span><span class="co">##  [53] lifecycle_1.0.3           yaml_2.3.7               </span></span>
<span><span class="co">##  [55] edgeR_3.41.6              carData_3.0-5            </span></span>
<span><span class="co">##  [57] rhdf5_2.43.4              blob_1.2.4               </span></span>
<span><span class="co">##  [59] grid_4.3.0                promises_1.2.0.1         </span></span>
<span><span class="co">##  [61] dqrng_0.3.0               crayon_1.5.2             </span></span>
<span><span class="co">##  [63] shinydashboard_0.7.2      lattice_0.21-8           </span></span>
<span><span class="co">##  [65] cowplot_1.1.1             beachmat_2.15.2          </span></span>
<span><span class="co">##  [67] annotate_1.77.0           KEGGREST_1.39.0          </span></span>
<span><span class="co">##  [69] magick_2.7.4              pillar_1.9.0             </span></span>
<span><span class="co">##  [71] knitr_1.42                rjson_0.2.21             </span></span>
<span><span class="co">##  [73] boot_1.3-28.1             fpc_2.2-10               </span></span>
<span><span class="co">##  [75] codetools_0.2-19          glue_1.6.2               </span></span>
<span><span class="co">##  [77] FCPS_1.3.1                data.table_1.14.8        </span></span>
<span><span class="co">##  [79] vctrs_0.6.1               png_0.1-8                </span></span>
<span><span class="co">##  [81] gtable_0.3.3              kernlab_0.9-32           </span></span>
<span><span class="co">##  [83] cachem_1.0.7              xfun_0.38                </span></span>
<span><span class="co">##  [85] princurve_2.1.6           mime_0.12                </span></span>
<span><span class="co">##  [87] DropletUtils_1.19.3       coop_0.6-3               </span></span>
<span><span class="co">##  [89] pheatmap_1.0.12           ellipsis_0.3.2           </span></span>
<span><span class="co">##  [91] nlme_3.1-162              bit64_4.0.5              </span></span>
<span><span class="co">##  [93] RcppAnnoy_0.0.20          rprojroot_2.0.3          </span></span>
<span><span class="co">##  [95] bslib_0.4.2               irlba_2.3.5.1            </span></span>
<span><span class="co">##  [97] svgPanZoom_0.3.4          vipor_0.4.5              </span></span>
<span><span class="co">##  [99] DBI_1.1.3                 colorspace_2.1-0         </span></span>
<span><span class="co">## [101] nnet_7.3-18               raster_3.6-20            </span></span>
<span><span class="co">## [103] mnormt_2.1.1              tidyselect_1.2.0         </span></span>
<span><span class="co">## [105] bit_4.0.5                 compiler_4.3.0           </span></span>
<span><span class="co">## [107] BiocNeighbors_1.17.1      desc_1.4.2               </span></span>
<span><span class="co">## [109] DelayedArray_0.25.0       bookdown_0.33            </span></span>
<span><span class="co">## [111] scales_1.2.1              DEoptimR_1.0-12          </span></span>
<span><span class="co">## [113] psych_2.3.3               tiff_0.1-11              </span></span>
<span><span class="co">## [115] stringr_1.5.0             SpatialExperiment_1.9.5  </span></span>
<span><span class="co">## [117] digest_0.6.31             goftest_1.2-3            </span></span>
<span><span class="co">## [119] fftwtools_0.9-11          spatstat.utils_3.0-2     </span></span>
<span><span class="co">## [121] minqa_1.2.5               rmarkdown_2.21           </span></span>
<span><span class="co">## [123] XVector_0.39.0            htmltools_0.5.5          </span></span>
<span><span class="co">## [125] pkgconfig_2.0.3           jpeg_0.1-10              </span></span>
<span><span class="co">## [127] lme4_1.1-32               sparseMatrixStats_1.11.1 </span></span>
<span><span class="co">## [129] highr_0.10                fastmap_1.1.1            </span></span>
<span><span class="co">## [131] rlang_1.1.0               htmlwidgets_1.6.2        </span></span>
<span><span class="co">## [133] shiny_1.7.4               DelayedMatrixStats_1.21.0</span></span>
<span><span class="co">## [135] farver_2.1.1              jquerylib_0.1.4          </span></span>
<span><span class="co">## [137] jsonlite_1.8.4            mclust_6.0.0             </span></span>
<span><span class="co">## [139] R.oo_1.25.0               BiocSingular_1.15.0      </span></span>
<span><span class="co">## [141] RCurl_1.98-1.12           magrittr_2.0.3           </span></span>
<span><span class="co">## [143] modeltools_0.2-23         GenomeInfoDbData_1.2.10  </span></span>
<span><span class="co">## [145] ggplotify_0.1.0           Rhdf5lib_1.21.0          </span></span>
<span><span class="co">## [147] munsell_0.5.0             Rcpp_1.0.10              </span></span>
<span><span class="co">## [149] viridis_0.6.2             stringi_1.7.12           </span></span>
<span><span class="co">## [151] zlibbioc_1.45.0           MASS_7.3-58.4            </span></span>
<span><span class="co">## [153] flexmix_2.3-19            plyr_1.8.8               </span></span>
<span><span class="co">## [155] parallel_4.3.0            ggrepel_0.9.3            </span></span>
<span><span class="co">## [157] deldir_1.0-6              Biostrings_2.67.2        </span></span>
<span><span class="co">## [159] splines_4.3.0             tensor_1.5               </span></span>
<span><span class="co">## [161] locfit_1.5-9.7            fastcluster_1.2.3        </span></span>
<span><span class="co">## [163] spatstat.geom_3.1-0       ggsignif_0.6.4           </span></span>
<span><span class="co">## [165] reshape2_1.4.4            ScaledMatrix_1.7.0       </span></span>
<span><span class="co">## [167] XML_3.99-0.14             evaluate_0.20            </span></span>
<span><span class="co">## [169] BiocManager_1.30.20       nloptr_2.0.3             </span></span>
<span><span class="co">## [171] tweenr_2.0.2              httpuv_1.6.9             </span></span>
<span><span class="co">## [173] tidyr_1.3.0               purrr_1.0.1              </span></span>
<span><span class="co">## [175] polyclip_1.10-4           ggforce_0.4.1            </span></span>
<span><span class="co">## [177] rsvd_1.0.5                broom_1.0.4              </span></span>
<span><span class="co">## [179] xtable_1.8-4              rstatix_0.7.2            </span></span>
<span><span class="co">## [181] later_1.3.0               viridisLite_0.4.1        </span></span>
<span><span class="co">## [183] class_7.3-21              ragg_1.2.5               </span></span>
<span><span class="co">## [185] tibble_3.2.1              lmerTest_3.1-3           </span></span>
<span><span class="co">## [187] AnnotationDbi_1.61.2      memoise_2.0.1            </span></span>
<span><span class="co">## [189] beeswarm_0.4.0            cluster_2.1.4            </span></span>
<span><span class="co">## [191] concaveman_1.1.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Alexander Nicholls, Nicholas Robertson, Nicolas Canete, Elijah Willie, Ellis Patrick, SOMS Maintainer.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
